rust-install/src/component/package.rs

```diff
@@ -168,14 +168,35 @@ pub struct TarPackage<'a>(DirectoryPackage, temp::Dir<'a>);
 impl<'a> TarPackage<'a> {
     pub fn new<R: Read>(stream: R, temp_cfg: &'a temp::Cfg) -> Result<Self> {
         let temp_dir = try!(temp_cfg.new_directory());
-
         let mut archive = tar::Archive::new(stream);
-        try!(archive.unpack(&*temp_dir).map_err(Error::ExtractingPackage));
+        // The rust-installer packages unpack to a directory called
+        // $pkgname-$version-$target. Skip that directory when
+        // unpacking.
+        try!(unpack_without_first_dir(&mut archive, &*temp_dir));
 
         Ok(TarPackage(try!(DirectoryPackage::new(temp_dir.to_owned())), temp_dir))
     }
 }
 
+fn unpack_without_first_dir<R: Read>(archive: &mut tar::Archive<R>, path: &Path) -> Result<()> {
+    let entries = try!(archive.entries().map_err(Error::ExtractingPackage));
```

Comment:
Unrelated: this would be another use case for rust-lang/rfcs#243. You could write:

```rust
catch {
    let entries = archive.entries()?;
    ...
}.map_err(Error::ExtractingPackage)
```